ee/clickhouse/queries/actor_base_query.py

```diff
@@ -54,8 +54,16 @@ def __init__(
         self.entity = entity
         self._filter = filter
 
-    def actor_query(self) -> Tuple[str, Dict]:
-        """ Implemented by subclasses. Must provide query and params. The query must return list of uuids. Can be group uuids (group_key) or person uuids """
+    def actor_query(self) -> Tuple[str, Union[Dict, None]]:
+        """
+        Implemented by subclasses. Must provide query and params. The query must
+        return list of uuids. Can be group uuids (group_key) or person uuids
+
+        NOTE: we have the option of `None` as a params return type, which is to
+        highlight that the query should not have param substitution run on it.
```

Comment:
This kind of smells - we've traded off one tricky logic for another one and are now living in a world with both.

Why can't you `substitute_params(substitute_params(query))`?

Reply:
The issue with double substitute calls is that if you have substituted with a string that contains '%' then you're going to have issues if you then try to substitute again.

See the test for example of something that was failing btw: https://github.com/PostHog/posthog/pull/7748/files#diff-2295276f5c251ba546ce175b40ed781b6d2e74dff26d1a5f4a5c41af5a20c400R323

Not ideal being able to trip up over double substitutions so easily. I don't see a simple solution right now, aside from perhaps providing a helper that will handle 

  1. the param substitutions, and 
  2. the subquery/partial query substitutions. Perhaps something like:

```
render_query(
    query_template, 
    params={"offset": filter.offset}, 
    subqueries={"target_event_query": target_event_query}
)
```

This would at least handle the ordering of substitution consistently.

The alternative is perhaps we go back to gathering the substitution params. I wasn't keen on this as I found it difficult to understand what was going on in isolation and was bleeding out internal details into calling functions. I'm sure there's a nice simple way to do something nice here, I'm going to have a look at what sqlalchemy does with [bound params](https://docs.sqlalchemy.org/en/14/core/sqlelement.html#sqlalchemy.sql.expression.TextClause.bindparams) to see if there's a simple pattern we can use.

What does your gut tell you?

Reply:
Also, would you like me to resolve this as a requirement of this PR?

Reply:
> What does your gut tell you?

I would avoid adding a third solution (the combined fn) because we're just layering more abstractions at that point without getting closer to solving problems better. 

I'd say the blocker for me is the new return type being Optional[Dict] - it feels every callsite gets messier than it was before this way. 

I'd also put a pause on `substitute_params` and revisit it later to see if we can use it without that gotcha and then iterate from there.

Reply:
> I would avoid adding a third solution

Yes wasn't suggesting we use, just understand if there's a pattern we can lift.

> I'd say the blocker for me is the new return type being Optional[Dict]

Ok, I'll update this to avoid changing the type.

Reply:
@macobo I've made some changes to avoid the type changes, let me know if this is ok

I've also put up a little pr separately to highlight an alternative to avoid exposing the dangerous `substitute_params` function [here](https://github.com/PostHog/posthog/pull/7749/files)